File: Semester5/Parallel and Distributed Programming/Tema2/Java/src/SequentialMethod.java
public class SequentialMethod {
    private final int[][] matrix;
    private final int[][] kernel;
    private final int N;
    private final int M;
    private final int kernelSize;

    public SequentialMethod(int[][] matrix, int[][] kernel) {
        this.matrix = matrix;
        this.kernel = kernel;
        N = matrix.length;
        M = matrix[0].length;
        kernelSize = kernel.length;
    }

    public void start() {
        int[] previousRow = new int[M];
        int[] currentRow = new int[M];
        int[] nextRow = new int[M];

        System.arraycopy(matrix[0], 0, previousRow, 0, M);

        for (int i = 0; i < N; i++) {
            System.arraycopy(matrix[i], 0, currentRow, 0, M);

            if (i < N - 1) {
                System.arraycopy(matrix[i + 1], 0, nextRow, 0, M);
            }
            else {
                System.arraycopy(matrix[i], 0, nextRow, 0, M);
            }

            for (int j = 0; j < M; j++) {
                matrix[i][j] = Utils.convolveAt(previousRow, currentRow, nextRow, kernel, i, j, N, M, kernelSize);
            }

            System.arraycopy(currentRow, 0, previousRow, 0, M);
        }
    }
}

File: Semester5/Parallel and Distributed Programming/Tema2/C++/C++/SequentialMethod.cpp
#include "SequentialMethod.h"
#include <sstream>
#include <algorithm>
#include "Utils.h"

SequentialMethod::SequentialMethod(std::vector<std::vector<int>>& matrix, const std::vector<std::vector<int>>& kernel): matrix(matrix), kernel(kernel){}

void SequentialMethod::start()
{
    int N = matrix.size();
    int M = matrix[0].size();
    int K = kernel.size();

	std::vector<int> previousRow(M);
	std::vector<int> currentRow(M);
	std::vector<int> nextRow(M);

	std::copy(matrix[0].begin(), matrix[0].end(), previousRow.begin());

    for (int i = 0; i < N; i++) {
		std::copy(matrix[i].begin(), matrix[i].end(), currentRow.begin());

        if (i < N - 1) {
			std::copy(matrix[i + 1].begin(), matrix[i + 1].end(), nextRow.begin());
        }
        else {
			std::copy(matrix[i].begin(), matrix[i].end(), nextRow.begin());
        }

        for (int j = 0; j < M; j++) {
			matrix[i][j] = Utils::convolveAt(previousRow, currentRow, nextRow, kernel, i, j, N, M);
        }

		std::copy(currentRow.begin(), currentRow.end(), previousRow.begin());
    }
}
